### DISCUSSION TITLE
Allow with operator to be used with non-record/struct types

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#8944] [Language Ideas]
If I have a struct or a record, I can do this:

```
var someObject = someOtherObject with { SomeProperty = "some value" };
```

However this doesn't work if `someOtherObject` is, say, a class type, instead of a record or struct type. This seems like an arbitrary restriction and it would be useful to be able to use the `with` operator with other types such as classes.

**Comment by HaloFour** [comment]
See #8876

There are technical reasons for the limitation, namely that unless the type adheres strictly to the pattern that it will result in decapitation.  The compiler ensures that records adhere to that pattern for you.

```cs
public class Person {
    public string Name { get; init; }

    public virtual Person Clone() { ... }
}

public class Student : Person {
    public string Class { get; init; }

    // forgot to override Clone
}

Person person = new Student() { Name = "Foo", Class = "Underwater Basket Weaving" };

var cloned = person with { Name = "Bar" };

Debug.Assert(cloned is Student); // oops, not a Student!
```

**Comment by 333fred** [comment]
As Halo pointed out, this is a duplicate of https://github.com/dotnet/csharplang/discussions/8876. Closing as such.
